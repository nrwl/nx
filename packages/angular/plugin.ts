export { createNodesV2 } from './src/plugins/plugin';
