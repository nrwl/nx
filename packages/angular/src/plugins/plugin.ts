import {
  type CreateNodesContextV2,
  createNodesFromFiles,
  type CreateNodesResult,
  type CreateNodesV2,
  detectPackageManager,
  getPackageManagerCommand,
  type ProjectConfiguration,
  readJsonFile,
  type Target,
  type TargetConfiguration,
  writeJsonFile,
} from '@nx/devkit';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { getLockFileName } from '@nx/js';
import { existsSync, readdirSync, statSync } from 'node:fs';
import { dirname, join, relative } from 'node:path';
import * as posix from 'node:path/posix';
import { hashObject } from 'nx/src/devkit-internals';
import { workspaceDataDirectory } from 'nx/src/utils/cache-directory';

export interface AngularPluginOptions {
  targetNamePrefix?: string;
}

type AngularProjects = Record<
  string,
  Pick<ProjectConfiguration, 'projectType' | 'sourceRoot' | 'targets'>
>;

type AngularTargetConfiguration = {
  builder: string;
  options?: Record<string, any>;
  configurations?: Record<string, any>;
  defaultConfiguration?: string;
};
export type AngularProjectConfiguration = {
  projectType: 'application' | 'library';
  root: string;
  sourceRoot?: string;
  architect?: Record<string, AngularTargetConfiguration>;
  targets?: Record<string, AngularTargetConfiguration>;
};
type AngularJson = { projects?: Record<string, AngularProjectConfiguration> };

const knownExecutors = {
  appShell: new Set(['@angular-devkit/build-angular:app-shell']),
  build: new Set([
    '@angular-devkit/build-angular:application',
    '@angular/build:application',
    '@angular-devkit/build-angular:browser-esbuild',
    '@angular-devkit/build-angular:browser',
    '@angular-devkit/build-angular:ng-packagr',
    '@angular/build:ng-packagr',
  ]),
  devServer: new Set(['@angular-devkit/build-angular:dev-server']),
  extractI18n: new Set(['@angular-devkit/build-angular:extract-i18n']),
  prerender: new Set([
    '@angular-devkit/build-angular:prerender',
    '@nguniversal/builders:prerender',
  ]),
  server: new Set(['@angular-devkit/build-angular:server']),
  serveSsr: new Set([
    '@angular-devkit/build-angular:ssr-dev-server',
    '@nguniversal/builders:ssr-dev-server',
  ]),
  test: new Set(['@angular-devkit/build-angular:karma']),
};

const pmc = getPackageManagerCommand();

function readProjectsCache(cachePath: string): Record<string, AngularProjects> {
  return existsSync(cachePath) ? readJsonFile(cachePath) : {};
}

function writeProjectsToCache(
  cachePath: string,
  results: Record<string, AngularProjects>
) {
  writeJsonFile(cachePath, results);
}

export const createNodesV2: CreateNodesV2<AngularPluginOptions> = [
  '**/angular.json',
  async (configFiles, options, context) => {
    const optionsHash = hashObject(options);
    const cachePath = join(
      workspaceDataDirectory,
      `angular-${optionsHash}.hash`
    );
    const projectsCache = readProjectsCache(cachePath);
    try {
      return await createNodesFromFiles(
        (configFile, options, context) =>
          createNodesInternal(configFile, options, context, projectsCache),
        configFiles,
        options,
        context
      );
    } finally {
      writeProjectsToCache(cachePath, projectsCache);
    }
  },
];

async function createNodesInternal(
  configFilePath: string,
  options: {} | undefined,
  context: CreateNodesContextV2,
  projectsCache: Record<string, AngularProjects>
): Promise<CreateNodesResult> {
  const angularWorkspaceRoot = dirname(configFilePath);

  // Do not create a project if package.json isn't there
  const siblingFiles = readdirSync(
    join(context.workspaceRoot, angularWorkspaceRoot)
  );
  if (!siblingFiles.includes('package.json')) {
    return {};
  }

  const hash = await calculateHashForCreateNodes(
    angularWorkspaceRoot,
    options,
    context,
    [getLockFileName(detectPackageManager(context.workspaceRoot))]
  );

  projectsCache[hash] ??= await buildAngularProjects(
    configFilePath,
    options,
    angularWorkspaceRoot,
    context
  );

  return { projects: projectsCache[hash] };
}

async function buildAngularProjects(
  configFilePath: string,
  options: AngularPluginOptions,
  angularWorkspaceRoot: string,
  context: CreateNodesContextV2
): Promise<AngularProjects> {
  const projects: Record<string, AngularProjects[string] & { root: string }> =
    {};

  const absoluteConfigFilePath = join(context.workspaceRoot, configFilePath);
  const angularJson = readJsonFile<AngularJson>(absoluteConfigFilePath);

  const appShellTargets: Target[] = [];
  const prerenderTargets: Target[] = [];
  for (const [projectName, project] of Object.entries(
    angularJson.projects ?? {}
  )) {
    const targets: Record<string, TargetConfiguration> = {};

    const projectTargets = getAngularJsonProjectTargets(project);
    if (!projectTargets) {
      continue;
    }

    const namedInputs = getNamedInputs(project.root, context);

    for (const [angularTargetName, angularTarget] of Object.entries(
      projectTargets
    )) {
      const nxTargetName = options?.targetNamePrefix
        ? `${options.targetNamePrefix}${angularTargetName}`
        : angularTargetName;
      const externalDependencies = ['@angular/cli'];

      targets[nxTargetName] = {
        command:
          // For targets that are also Angular CLI commands, infer the simplified form.
          // Otherwise, use `ng run` to support non-command targets so that they will run.
          angularTargetName === 'build' ||
          angularTargetName === 'deploy' ||
          angularTargetName === 'extract-i18n' ||
          angularTargetName === 'e2e' ||
          angularTargetName === 'lint' ||
          angularTargetName === 'serve' ||
          angularTargetName === 'test'
            ? `ng ${angularTargetName}`
            : `ng run ${projectName}:${angularTargetName}`,
        options: { cwd: angularWorkspaceRoot },
        metadata: {
          technologies: ['angular'],
          description: `Run the "${angularTargetName}" target for "${projectName}".`,
          help: {
            command: `${pmc.exec} ng run ${projectName}:${angularTargetName} --help`,
            example: {},
          },
        },
      };

      if (knownExecutors.appShell.has(angularTarget.builder)) {
        appShellTargets.push({ target: nxTargetName, project: projectName });
      } else if (knownExecutors.build.has(angularTarget.builder)) {
        await updateBuildTarget(
          nxTargetName,
          targets[nxTargetName],
          angularTarget,
          context,
          angularWorkspaceRoot,
          project.root,
          namedInputs
        );
      } else if (knownExecutors.devServer.has(angularTarget.builder)) {
        targets[nxTargetName].metadata.help.example.options = { port: 4201 };
      } else if (knownExecutors.extractI18n.has(angularTarget.builder)) {
        targets[nxTargetName].metadata.help.example.options = {
          format: 'json',
        };
      } else if (knownExecutors.test.has(angularTarget.builder)) {
        updateTestTarget(
          targets[nxTargetName],
          angularTarget,
          context,
          angularWorkspaceRoot,
          project.root,
          namedInputs,
          externalDependencies
        );
      } else if (knownExecutors.server.has(angularTarget.builder)) {
        updateServerTarget(
          targets[nxTargetName],
          angularTarget,
          context,
          angularWorkspaceRoot,
          project.root,
          namedInputs
        );
      } else if (knownExecutors.serveSsr.has(angularTarget.builder)) {
        targets[nxTargetName].metadata.help.example.options = { port: 4201 };
      } else if (knownExecutors.prerender.has(angularTarget.builder)) {
        prerenderTargets.push({ target: nxTargetName, project: projectName });
      }

      if (targets[nxTargetName].inputs?.length) {
        targets[nxTargetName].inputs.push({ externalDependencies });
      }

      if (angularTarget.configurations) {
        for (const configurationName of Object.keys(
          angularTarget.configurations
        )) {
          targets[nxTargetName].configurations = {
            ...targets[nxTargetName].configurations,
            [configurationName]: {
              command: `ng run ${projectName}:${angularTargetName}:${configurationName}`,
            },
          };
        }
      }

      if (angularTarget.defaultConfiguration) {
        targets[nxTargetName].defaultConfiguration =
          angularTarget.defaultConfiguration;
      }
    }

    projects[projectName] = {
      projectType: project.projectType,
      root: posix.join(angularWorkspaceRoot, project.root),
      sourceRoot: project.sourceRoot
        ? posix.join(angularWorkspaceRoot, project.sourceRoot)
        : undefined,
      targets,
    };
  }

  for (const { project, target } of appShellTargets) {
    updateAppShellTarget(
      project,
      target,
      projects,
      angularJson,
      angularWorkspaceRoot,
      context
    );
  }

  for (const { project, target } of prerenderTargets) {
    updatePrerenderTarget(project, target, projects, angularJson);
  }

  return Object.entries(projects).reduce((acc, [projectName, project]) => {
    acc[project.root] = {
      projectType: project.projectType,
      sourceRoot: project.sourceRoot,
      targets: project.targets,
    };
    return acc;
  }, {} as AngularProjects);
}

function updateAppShellTarget(
  projectName: string,
  targetName: string,
  projects: AngularProjects,
  angularJson: AngularJson,
  angularWorkspaceRoot: string,
  context: CreateNodesContextV2
): void {
  // it must exist since we collected it when processing it
  const target = projects[projectName].targets[targetName];

  target.metadata.help.example.options = { route: '/some/route' };

  const { inputs, outputs } = getBrowserAndServerTargetInputsAndOutputs(
    projectName,
    targetName,
    projects,
    angularJson
  );

  const outputIndexPath = getAngularJsonProjectTargets(
    angularJson.projects[projectName]
  )[targetName].options?.outputIndexPath;
  if (outputIndexPath) {
    const fullOutputIndexPath = join(
      context.workspaceRoot,
      angularWorkspaceRoot,
      outputIndexPath
    );
    outputs.push(
      getOutput(
        fullOutputIndexPath,
        context.workspaceRoot,
        angularWorkspaceRoot,
        angularJson.projects[projectName].root
      )
    );
  }

  if (!outputs.length) {
    // no outputs were identified for the build or server target, so we don't
    // set any Nx cache options
    return;
  }

  target.cache = true;
  target.inputs = inputs;
  target.outputs = outputs;
}

async function updateBuildTarget(
  targetName: string,
  target: TargetConfiguration,
  angularTarget: AngularTargetConfiguration,
  context: CreateNodesContextV2,
  angularWorkspaceRoot: string,
  projectRoot: string,
  namedInputs: ReturnType<typeof getNamedInputs>
): Promise<void> {
  target.dependsOn = [`^${targetName}`];

  if (angularTarget.options?.outputPath) {
    const fullOutputPath = join(
      context.workspaceRoot,
      angularWorkspaceRoot,
      angularTarget.options.outputPath
    );
    target.outputs = [
      getOutput(
        fullOutputPath,
        context.workspaceRoot,
        angularWorkspaceRoot,
        projectRoot
      ),
    ];
  } else if (
    angularTarget.builder === '@angular-devkit/build-angular:ng-packagr' ||
    angularTarget.builder === '@angular/build:ng-packagr'
  ) {
    const outputs = await getNgPackagrOutputs(
      angularTarget,
      angularWorkspaceRoot,
      projectRoot,
      context
    );
    if (outputs.length) {
      target.outputs = outputs;
    }
  }

  if (target.outputs?.length) {
    // make it cacheable if we were able to identify outputs
    target.cache = true;
    target.inputs =
      'production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default'];
  }

  if (
    angularTarget.builder === '@angular-devkit/build-angular:ng-packagr' ||
    angularTarget.builder === '@angular/build:ng-packagr'
  ) {
    target.metadata.help.example.options = { watch: true };
  } else {
    target.metadata.help.example.options = { localize: true };
  }
}

function updateTestTarget(
  target: TargetConfiguration,
  angularTarget: AngularTargetConfiguration,
  context: CreateNodesContextV2,
  angularWorkspaceRoot: string,
  projectRoot: string,
  namedInputs: ReturnType<typeof getNamedInputs>,
  externalDependencies: string[]
): void {
  target.cache = true;
  target.inputs =
    'production' in namedInputs
      ? ['default', '^production']
      : ['default', '^default'];
  target.outputs = getKarmaTargetOutputs(
    angularTarget,
    angularWorkspaceRoot,
    projectRoot,
    context
  );
  externalDependencies.push('karma');

  target.metadata.help.example.options = { codeCoverage: true };
}

function updateServerTarget(
  target: TargetConfiguration,
  angularTarget: AngularTargetConfiguration,
  context: CreateNodesContextV2,
  angularWorkspaceRoot: string,
  projectRoot: string,
  namedInputs: ReturnType<typeof getNamedInputs>
): void {
  target.metadata.help.example.options = { localize: true };

  if (!angularTarget.options?.outputPath) {
    // only make it cacheable if we were able to identify outputs
    return;
  }

  target.cache = true;
  target.inputs =
    'production' in namedInputs
      ? ['production', '^production']
      : ['default', '^default'];

  const fullOutputPath = join(
    context.workspaceRoot,
    angularWorkspaceRoot,
    angularTarget.options.outputPath
  );
  target.outputs = [
    getOutput(
      fullOutputPath,
      context.workspaceRoot,
      angularWorkspaceRoot,
      projectRoot
    ),
  ];
}

function updatePrerenderTarget(
  projectName: string,
  targetName: string,
  projects: AngularProjects,
  angularJson: AngularJson
): void {
  // it must exist since we collected it when processing it
  const target = projects[projectName].targets[targetName];

  target.metadata.help.example.options =
    getAngularJsonProjectTargets(angularJson.projects[projectName])[targetName]
      .builder === '@angular-devkit/build-angular:prerender'
      ? { discoverRoutes: false }
      : { guessRoutes: false };

  const { inputs, outputs } = getBrowserAndServerTargetInputsAndOutputs(
    projectName,
    targetName,
    projects,
    angularJson
  );

  if (!outputs.length) {
    // no outputs were identified for the build or server target, so we don't
    // set any Nx cache options
    return;
  }

  target.cache = true;
  target.inputs = inputs;
  target.outputs = outputs;
}

async function getNgPackagrOutputs(
  target: AngularTargetConfiguration,
  angularWorkspaceRoot: string,
  projectRoot: string,
  context: CreateNodesContextV2
): Promise<string[]> {
  let ngPackageJsonPath = join(
    context.workspaceRoot,
    angularWorkspaceRoot,
    target.options.project
  );

  const readConfig = async (configPath: string) => {
    if (!existsSync(configPath)) {
      return undefined;
    }

    try {
      if (configPath.endsWith('.js')) {
        const result = await import(configPath);

        return result['default'] ?? result;
      }

      return readJsonFile(configPath);
    } catch {}

    return undefined;
  };

  let ngPackageJson: { dest?: string };
  let basePath: string;
  if (statSync(ngPackageJsonPath).isDirectory()) {
    basePath = ngPackageJsonPath;
    ngPackageJson = await readConfig(
      join(ngPackageJsonPath, 'ng-package.json')
    );
    if (!ngPackageJson) {
      ngPackageJson = await readConfig(
        join(ngPackageJsonPath, 'ng-package.js')
      );
    }
  } else {
    basePath = dirname(ngPackageJsonPath);
    ngPackageJson = await readConfig(ngPackageJsonPath);
  }

  if (!ngPackageJson) {
    return [];
  }

  const destination = ngPackageJson.dest
    ? join(basePath, ngPackageJson.dest)
    : join(basePath, 'dist');

  return [
    getOutput(
      destination,
      context.workspaceRoot,
      angularWorkspaceRoot,
      projectRoot
    ),
  ];
}

function getKarmaTargetOutputs(
  target: AngularTargetConfiguration,
  angularWorkspaceRoot: string,
  projectRoot: string,
  context: CreateNodesContextV2
): string[] {
  const defaultOutput = posix.join(
    '{workspaceRoot}',
    angularWorkspaceRoot,
    'coverage/{projectName}'
  );
  if (!target.options?.karmaConfig) {
    return [defaultOutput];
  }

  try {
    const { parseConfig } = require('karma/lib/config');

    const karmaConfigPath = join(
      context.workspaceRoot,
      angularWorkspaceRoot,
      projectRoot,
      target.options.karmaConfig
    );
    const config = parseConfig(karmaConfigPath);

    if (config.coverageReporter.dir) {
      return [
        getOutput(
          config.coverageReporter.dir,
          context.workspaceRoot,
          angularWorkspaceRoot,
          projectRoot
        ),
      ];
    }
  } catch {
    // we silently ignore any error here and fall back to the default output
  }

  return [defaultOutput];
}

function getBrowserAndServerTargetInputsAndOutputs(
  projectName: string,
  targetName: string,
  projects: AngularProjects,
  angularJson: AngularJson
) {
  const { browserTarget, serverTarget } = extractBrowserAndServerTargets(
    angularJson,
    projectName,
    targetName
  );
  if (!browserTarget || !serverTarget) {
    // if any of these are missing, the target is invalid so we return empty values
    return { inputs: [], outputs: [] };
  }

  const browserTargetInputs =
    projects[browserTarget.project]?.targets?.[browserTarget.target]?.inputs ??
    [];
  const serverTargetInputs =
    projects[serverTarget.project]?.targets?.[serverTarget.target]?.inputs ??
    [];
  const browserTargetOutputs =
    projects[browserTarget.project]?.targets?.[browserTarget.target]?.outputs ??
    [];
  const serverTargetOutputs =
    projects[serverTarget.project]?.targets?.[serverTarget.target]?.outputs ??
    [];

  return {
    inputs: mergeInputs(...browserTargetInputs, ...serverTargetInputs),
    outputs: Array.from(
      new Set([...browserTargetOutputs, ...serverTargetOutputs])
    ),
  };
}

function extractBrowserAndServerTargets(
  angularJson: AngularJson,
  projectName: string,
  targetName: string
): {
  browserTarget: Target;
  serverTarget: Target;
} {
  let browserTarget: Target | undefined;
  let serverTarget: Target | undefined;

  try {
    const targets = getAngularJsonProjectTargets(
      angularJson.projects[projectName]
    );
    const target = targets[targetName];

    let browserTargetSpecifier = target.options?.browserTarget;
    if (!browserTargetSpecifier) {
      const configuration = Object.values(target.configurations ?? {}).find(
        (config) => !!config.browserTarget
      );
      browserTargetSpecifier = configuration?.browserTarget;
    }

    if (browserTargetSpecifier) {
      browserTarget = targetFromTargetString(
        browserTargetSpecifier,
        projectName,
        targetName
      );
    }

    let serverTargetSpecifier = target.options?.serverTarget;
    if (!serverTargetSpecifier) {
      serverTargetSpecifier = Object.values(target.configurations ?? {}).find(
        (config) => !!config.serverTarget
      )?.serverTarget;
    }

    if (serverTargetSpecifier) {
      serverTarget = targetFromTargetString(
        serverTargetSpecifier,
        projectName,
        targetName
      );
    }
  } catch {}

  return { browserTarget: browserTarget, serverTarget };
}

function mergeInputs(
  ...inputs: TargetConfiguration['inputs']
): TargetConfiguration['inputs'] {
  const stringInputs = new Set<string>();
  const externalDependencies = new Set<string>();

  for (const input of inputs) {
    if (typeof input === 'string') {
      stringInputs.add(input);
    } else if ('externalDependencies' in input) {
      // we only infer external dependencies, so we don't need to handle the other input definitions
      for (const externalDependency of input.externalDependencies) {
        externalDependencies.add(externalDependency);
      }
    }
  }

  return [
    ...stringInputs,
    ...(externalDependencies.size
      ? [{ externalDependencies: Array.from(externalDependencies) }]
      : []),
  ];
}

// angular support abbreviated target specifiers, this is adapter from:
// https://github.com/angular/angular-cli/blob/7d9ce246a33c60ec96eb4bf99520f5475716a910/packages/angular_devkit/architect/src/api.ts#L336
function targetFromTargetString(
  specifier: string,
  abbreviatedProjectName?: string,
  abbreviatedTargetName?: string
) {
  const tuple = specifier.split(':', 3);
  if (tuple.length < 2) {
    // invalid target, ignore
    return undefined;
  }

  // we only care about project and target
  return {
    project: tuple[0] || abbreviatedProjectName || '',
    target: tuple[1] || abbreviatedTargetName || '',
  };
}

function getOutput(
  path: string,
  workspaceRoot: string,
  angularWorkspaceRoot: string,
  projectRoot: string
): string {
  const relativePath = relative(
    join(workspaceRoot, angularWorkspaceRoot, projectRoot),
    path
  );
  if (relativePath.startsWith('..')) {
    return posix.join(
      '{workspaceRoot}',
      join(angularWorkspaceRoot, projectRoot, relativePath)
    );
  } else {
    return posix.join('{projectRoot}', relativePath);
  }
}

function getAngularJsonProjectTargets(
  project: AngularProjectConfiguration
): Record<string, AngularTargetConfiguration> {
  return project.architect ?? project.targets;
}
