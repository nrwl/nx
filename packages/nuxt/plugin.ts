export {
  createNodes,
  createNodesV2,
  NuxtPluginOptions,
} from './src/plugins/plugin';
