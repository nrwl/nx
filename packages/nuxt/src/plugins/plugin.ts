import type { NuxtOptions } from '@nuxt/schema';
import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  readJsonFile,
  TargetConfiguration,
  workspaceRoot,
  writeJsonFile,
} from '@nx/devkit';
import { loadConfigFile } from '@nx/devkit/src/utils/config-utils';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { workspaceDataDirectory } from 'nx/src/utils/cache-directory';
import { getLockFileName } from '@nx/js';
import { dirname, isAbsolute, join, relative } from 'path';
import { existsSync, readdirSync } from 'fs';
import { loadNuxtKitDynamicImport } from '../utils/executor-utils';

const cachePath = join(workspaceDataDirectory, 'nuxt.hash');
const targetsCache = readTargetsCache();

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration>
> {
  return existsSync(cachePath) ? readJsonFile(cachePath) : {};
}

function writeTargetsToCache() {
  const oldCache = readTargetsCache();
  writeJsonFile(cachePath, {
    ...oldCache,
    ...targetsCache,
  });
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache();
  return [];
};

export interface NuxtPluginOptions {
  buildTargetName?: string;
  serveTargetName?: string;
  serveStaticTargetName?: string;
  buildStaticTargetName?: string;
}

export const createNodes: CreateNodes<NuxtPluginOptions> = [
  '**/nuxt.config.{js,ts,mjs,mts,cjs,cts}',
  async (configFilePath, options, context) => {
    const projectRoot = dirname(configFilePath);
    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    options = normalizeOptions(options);

    const hash = await calculateHashForCreateNodes(
      projectRoot,
      options,
      context,
      [getLockFileName(detectPackageManager(context.workspaceRoot))]
    );
    targetsCache[hash] ??= await buildNuxtTargets(
      configFilePath,
      projectRoot,
      options,
      context
    );

    return {
      projects: {
        [projectRoot]: {
          root: projectRoot,
          targets: targetsCache[hash],
        },
      },
    };
  },
];

async function buildNuxtTargets(
  configFilePath: string,
  projectRoot: string,
  options: NuxtPluginOptions,
  context: CreateNodesContext
) {
  const nuxtConfig: {
    buildDir: string;
  } = await getInfoFromNuxtConfig(configFilePath, context, projectRoot);

  const { buildOutputs } = getOutputs(nuxtConfig, projectRoot);

  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {};

  targets[options.buildTargetName] = buildTarget(
    options.buildTargetName,
    namedInputs,
    buildOutputs,
    projectRoot
  );

  targets[options.serveTargetName] = serveTarget(projectRoot);

  targets[options.serveStaticTargetName] = serveStaticTarget(options);

  targets[options.buildStaticTargetName] = buildStaticTarget(
    options.buildStaticTargetName,
    namedInputs,
    buildOutputs,
    projectRoot
  );

  return targets;
}

function buildTarget(
  buildTargetName: string,
  namedInputs: {
    [inputName: string]: any[];
  },
  buildOutputs: string[],
  projectRoot: string
) {
  return {
    command: `nuxt build`,
    options: { cwd: projectRoot },
    cache: true,
    dependsOn: [`^${buildTargetName}`],
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),

      {
        externalDependencies: ['nuxt'],
      },
    ],
    outputs: buildOutputs,
  };
}

function serveTarget(projectRoot: string) {
  const targetConfig: TargetConfiguration = {
    command: `nuxt dev`,
    options: {
      cwd: projectRoot,
    },
  };

  return targetConfig;
}

function serveStaticTarget(options: NuxtPluginOptions) {
  const targetConfig: TargetConfiguration = {
    executor: '@nx/web:file-server',
    options: {
      buildTarget: `${options.buildStaticTargetName}`,
      staticFilePath: '{projectRoot}/dist',
      port: 4200,
      // Routes are found correctly with serve-static
      spa: false,
    },
  };

  return targetConfig;
}

function buildStaticTarget(
  buildStaticTargetName: string,
  namedInputs: {
    [inputName: string]: any[];
  },
  buildOutputs: string[],
  projectRoot: string
) {
  const targetConfig: TargetConfiguration = {
    command: `nuxt build --prerender`,
    options: { cwd: projectRoot },
    cache: true,
    dependsOn: [`^${buildStaticTargetName}`],
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),

      {
        externalDependencies: ['nuxt'],
      },
    ],
    outputs: buildOutputs,
  };
  return targetConfig;
}

async function getInfoFromNuxtConfig(
  configFilePath: string,
  context: CreateNodesContext,
  projectRoot: string
): Promise<{
  buildDir: string;
}> {
  let config: NuxtOptions;
  if (process.env.NX_ISOLATE_PLUGINS !== 'false') {
    config = await (
      await loadNuxtKitDynamicImport()
    ).loadNuxtConfig({
      configFile: configFilePath,
    });
  } else {
    config = await loadConfigFile(join(context.workspaceRoot, configFilePath));
  }
  return {
    buildDir:
      config?.buildDir ??
      // Match .nuxt default build dir from '@nuxt/schema'
      // See: https://github.com/nuxt/nuxt/blob/871404ae5673425aeedde82f123ea58aa7c6facf/packages/schema/src/config/common.ts#L117-L119
      '.nuxt',
  };
}

function getOutputs(
  nuxtConfig: { buildDir: string },
  projectRoot: string
): {
  buildOutputs: string[];
} {
  const buildOutputPath = normalizeOutputPath(
    nuxtConfig?.buildDir,
    projectRoot
  );

  return {
    buildOutputs: [buildOutputPath, '{projectRoot}/.output'],
  };
}

function normalizeOutputPath(
  outputPath: string | undefined,
  projectRoot: string
): string {
  if (!outputPath) {
    if (projectRoot === '.') {
      return `{projectRoot}`;
    } else {
      return `{workspaceRoot}/{projectRoot}`;
    }
  } else {
    if (isAbsolute(outputPath)) {
      return `{workspaceRoot}/${relative(workspaceRoot, outputPath)}`;
    } else {
      if (outputPath.startsWith('..')) {
        return join('{workspaceRoot}', join(projectRoot, outputPath));
      } else {
        return join('{projectRoot}', outputPath);
      }
    }
  }
}

function normalizeOptions(options: NuxtPluginOptions): NuxtPluginOptions {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.serveTargetName ??= 'serve';
  options.serveStaticTargetName ??= 'serve-static';
  options.buildStaticTargetName ??= 'build-static';
  return options;
}
