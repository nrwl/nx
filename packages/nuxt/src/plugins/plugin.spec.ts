import { CreateNodesContextV2 } from '@nx/devkit';
import { createNodes } from './plugin';
import { TempFs } from 'nx/src/internal-testing-utils/temp-fs';

jest.mock('@nx/devkit/src/utils/config-utils', () => ({
  loadConfigFile: jest.fn().mockImplementation(() => {
    return Promise.resolve({
      buildDir: '../dist/my-app/.nuxt',
    });
  }),
}));

jest.mock('../utils/executor-utils', () => ({
  loadNuxtKitDynamicImport: jest.fn().mockResolvedValue({
    loadNuxtConfig: jest.fn().mockResolvedValue({
      buildDir: '../dist/my-app/.nuxt',
    }),
  }),
}));
describe('@nx/nuxt/plugin', () => {
  let createNodesFunction = createNodes[1];
  let context: CreateNodesContextV2;

  describe('root project', () => {
    beforeEach(async () => {
      context = {
        nxJsonConfiguration: {
          // These defaults should be overridden by plugin
          targetDefaults: {
            build: {
              cache: false,
              inputs: ['foo', '^foo'],
            },
          },
          namedInputs: {
            default: ['{projectRoot}/**/*'],
            production: ['!{projectRoot}/**/*.spec.ts'],
          },
        },
        workspaceRoot: '',
      };
    });

    afterEach(() => {
      jest.resetModules();
    });

    it('should create nodes', async () => {
      const nodes = await createNodesFunction(
        ['nuxt.config.ts'],
        {
          buildTargetName: 'build',
          serveTargetName: 'serve',
          buildStaticTargetName: 'build-static',
          serveStaticTargetName: 'serve-static',
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });
  });

  describe('not root project', () => {
    const tempFs = new TempFs('test');
    beforeEach(() => {
      context = {
        nxJsonConfiguration: {
          namedInputs: {
            default: ['{projectRoot}/**/*'],
            production: ['!{projectRoot}/**/*.spec.ts'],
          },
        },
        workspaceRoot: tempFs.tempDir,
      };

      tempFs.createFileSync(
        'my-app/project.json',
        JSON.stringify({ name: 'my-app' })
      );
      tempFs.createFileSync('my-app/nuxt.config.ts', '');
    });

    afterEach(() => {
      jest.resetModules();
    });

    it('should create nodes', async () => {
      const nodes = await createNodesFunction(
        ['my-app/nuxt.config.ts'],
        {
          buildTargetName: 'build-something',
          serveTargetName: 'my-serve',
          buildStaticTargetName: 'acme-build-static',
          serveStaticTargetName: 'acme-serve-static',
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });
  });
});
