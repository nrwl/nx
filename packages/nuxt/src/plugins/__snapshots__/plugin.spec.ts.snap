// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx/nuxt/plugin not root project should create nodes 1`] = `
{
  "projects": {
    "my-app": {
      "root": "my-app",
      "targets": {
        "acme-build-static": {
          "cache": true,
          "command": "nuxt build --prerender",
          "dependsOn": [
            "^acme-build-static",
          ],
          "inputs": [
            "production",
            "^production",
            {
              "externalDependencies": [
                "nuxt",
              ],
            },
          ],
          "options": {
            "cwd": "my-app",
          },
          "outputs": [
            "{workspaceRoot}/dist/my-app/.nuxt",
            "{projectRoot}/.output",
          ],
        },
        "acme-serve-static": {
          "executor": "@nx/web:file-server",
          "options": {
            "buildTarget": "acme-build-static",
            "port": 4200,
            "spa": false,
            "staticFilePath": "{projectRoot}/dist",
          },
        },
        "build-deps": {
          "dependsOn": [
            "^build",
          ],
        },
        "build-something": {
          "cache": true,
          "command": "nuxt build",
          "dependsOn": [
            "^build-something",
          ],
          "inputs": [
            "production",
            "^production",
            {
              "externalDependencies": [
                "nuxt",
              ],
            },
          ],
          "options": {
            "cwd": "my-app",
          },
          "outputs": [
            "{workspaceRoot}/dist/my-app/.nuxt",
            "{projectRoot}/.output",
          ],
        },
        "my-serve": {
          "command": "nuxt dev",
          "options": {
            "cwd": "my-app",
          },
        },
        "watch-deps": {
          "command": "npx nx watch --projects my-app --includeDependentProjects -- npx nx build-deps my-app",
          "dependsOn": [
            "build-deps",
          ],
        },
      },
    },
  },
}
`;

exports[`@nx/nuxt/plugin root project should create nodes 1`] = `
{
  "projects": {
    ".": {
      "root": ".",
      "targets": {
        "build": {
          "cache": true,
          "command": "nuxt build",
          "dependsOn": [
            "^build",
          ],
          "inputs": [
            "production",
            "^production",
            {
              "externalDependencies": [
                "nuxt",
              ],
            },
          ],
          "options": {
            "cwd": ".",
          },
          "outputs": [
            "dist/my-app/.nuxt",
            "{projectRoot}/.output",
          ],
        },
        "build-deps": {
          "dependsOn": [
            "^build",
          ],
        },
        "build-static": {
          "cache": true,
          "command": "nuxt build --prerender",
          "dependsOn": [
            "^build-static",
          ],
          "inputs": [
            "production",
            "^production",
            {
              "externalDependencies": [
                "nuxt",
              ],
            },
          ],
          "options": {
            "cwd": ".",
          },
          "outputs": [
            "dist/my-app/.nuxt",
            "{projectRoot}/.output",
          ],
        },
        "serve": {
          "command": "nuxt dev",
          "options": {
            "cwd": ".",
          },
        },
        "serve-static": {
          "executor": "@nx/web:file-server",
          "options": {
            "buildTarget": "build-static",
            "port": 4200,
            "spa": false,
            "staticFilePath": "{projectRoot}/dist",
          },
        },
        "watch-deps": {
          "command": "npx nx watch --projects nuxt --includeDependentProjects -- npx nx build-deps nuxt",
          "dependsOn": [
            "build-deps",
          ],
        },
      },
    },
  },
}
`;
